File: backend/services/logs.py
import json
from typing import Any
from uuid import UUID

import asyncpg


def _extract_trace(metadata: dict | None) -> dict[str, Any]:
    """從 messages.metadata 解析 tool calls 與 query rewrite。"""
    if not metadata or not isinstance(metadata, dict):
        return {"steps": [], "tool_calls": [], "query_rewrites": [], "error": None}

    steps = metadata.get("steps") or []
    tool_calls: list[dict] = []
    query_rewrites: list[dict] = []

    for step in steps:
        if not isinstance(step, dict):
            continue
        node = step.get("node", "unknown")
        for tc in step.get("tool_calls") or []:
            if isinstance(tc, dict):
                tool_calls.append({
                    "node": node,
                    "name": tc.get("name"),
                    "query": tc.get("query"),
                    "start_date": tc.get("start_date"),
                    "end_date": tc.get("end_date"),
                    "raw_args": tc.get("raw_args"),
                })

        extras = step.get("extras") or {}
        for key in ("flash_query_rewrite", "general_query_rewrite"):
            rewrite = extras.get(key)
            if isinstance(rewrite, dict):
                query_rewrites.append({
                    "node": node,
                    "type": key,
                    "pattern": rewrite.get("pattern"),
                    "rewritten_query": rewrite.get("rewritten_query"),
                    "original_query": rewrite.get("original_query"),
                })

    return {
        "steps": steps,
        "tool_calls": tool_calls,
        "query_rewrites": query_rewrites,
        "error": metadata.get("error"),
        "response_mode": metadata.get("response_mode"),
        "total_execution_time": metadata.get("total_execution_time"),
    }


async def list_chat_logs(
    db: asyncpg.Connection,
    *,
    user_id: UUID | None = None,
    chat_id: UUID | None = None,
    page: int = 1,
    limit: int = 20,
) -> dict[str, Any]:
    offset = (page - 1) * limit
    conditions = ["m.role = 'assistant'", "m.metadata IS NOT NULL"]
    params: list[Any] = []
    idx = 1

    if user_id:
        conditions.append(f"c.user_id = ${idx}")
        params.append(user_id)
        idx += 1

    if chat_id:
        conditions.append(f"m.chat_id = ${idx}")
        params.append(chat_id)
        idx += 1

    where = " AND ".join(conditions)

    total = await db.fetchval(
        f"""
        SELECT COUNT(*)
        FROM messages m
        JOIN chats c ON c.id = m.chat_id
        WHERE {where}
        """,
        *params,
    )

    rows = await db.fetch(
        f"""
        SELECT
            m.id AS message_id, m.chat_id, m.content,
            m.metadata, m.created_at,
            c.title AS chat_title, c.user_id,
            u.email, u.username
        FROM messages m
        JOIN chats c ON c.id = m.chat_id
        JOIN users u ON u.id = c.user_id
        WHERE {where}
        ORDER BY m.created_at DESC
        LIMIT ${idx} OFFSET ${idx + 1}
        """,
        *params,
        limit,
        offset,
    )

    items = []
    for row in rows:
        data = _serialize(row)
        meta = data.pop("metadata", None)
        if isinstance(meta, str):
            try:
                meta = json.loads(meta)
            except json.JSONDecodeError:
                meta = None
        trace = _extract_trace(meta)
        data["trace"] = trace
        data["tool_count"] = len(trace["tool_calls"])
        data["rewrite_count"] = len(trace["query_rewrites"])
        items.append(data)

    return {
        "total": total,
        "page": page,
        "limit": limit,
        "items": items,
    }


async def get_log_detail(
    db: asyncpg.Connection,
    message_id: UUID,
) -> dict | None:
    row = await db.fetchrow(
        """
        SELECT
            m.id AS message_id, m.chat_id, m.parent_id,
            m.role, m.content, m.tokens, m.context_refs,
            m.metadata, m.created_at,
            c.title AS chat_title, c.user_id,
            u.email, u.username
        FROM messages m
        JOIN chats c ON c.id = m.chat_id
        JOIN users u ON u.id = c.user_id
        WHERE m.id = $1
        """,
        message_id,
    )
    if not row:
        return None

    data = _serialize(row)
    for field in ("tokens", "context_refs", "metadata"):
        val = data.get(field)
        if isinstance(val, str):
            try:
                data[field] = json.loads(val)
            except json.JSONDecodeError:
                pass

    meta = data.get("metadata")
    data["trace"] = _extract_trace(meta if isinstance(meta, dict) else None)

    token_logs = await db.fetch(
        """
        SELECT id, caller, model_name, prompt_tokens, completion_tokens,
               total_tokens, cost_usd, created_at
        FROM token_usage_logs
        WHERE message_id = $1
        ORDER BY created_at ASC
        """,
        message_id,
    )
    data["token_logs"] = [_serialize(r) for r in token_logs]

    return data


def _serialize(row: asyncpg.Record) -> dict:
    result: dict[str, Any] = {}
    for key, val in dict(row).items():
        if isinstance(val, UUID):
            result[key] = str(val)
        elif hasattr(val, "isoformat"):
            result[key] = val.isoformat()
        else:
            result[key] = float(val) if hasattr(val, "as_tuple") else val
    return result

File: frontend/js/pages/logs.js
import { apiFetch, formatDate, truncate } from '../api.js';
import { el, clear, buildTable, buildPagination, showLoading, showError, createModal, badge, renderJson } from '../dom.js';
import { pageShell } from '../layout.js';

pageShell('聊天 Log', 'Tool 呼叫、Query Rewrite 與關鍵字追蹤', '/html/logs.html');

let currentPage = 1;
const limit = 20;
let userId = '';
let chatId = '';

const toolbar = document.getElementById('toolbar');
const tableContainer = document.getElementById('table-container');
const paginationContainer = document.getElementById('pagination-container');
const modalRoot = document.getElementById('modal-root');

const userInput = el('input', { className: 'input-field', placeholder: '使用者 UUID（可選）' });
const chatInput = el('input', { className: 'input-field', placeholder: '對話 UUID（可選）' });
const searchBtn = el('button', { className: 'btn btn-primary', textContent: '搜尋' });

searchBtn.addEventListener('click', () => {
    userId = userInput.value.trim();
    chatId = chatInput.value.trim();
    currentPage = 1;
    load();
});

toolbar.appendChild(userInput);
toolbar.appendChild(chatInput);
toolbar.appendChild(searchBtn);

async function load() {
    showLoading(tableContainer);
    clear(paginationContainer);
    try {
        const data = await apiFetch('/api/logs', {
            user_id: userId, chat_id: chatId, page: currentPage, limit,
        });

        clear(tableContainer);
        tableContainer.appendChild(buildTable([
            { label: '使用者', render: r => r.email || r.username },
            { label: '對話', render: r => truncate(r.chat_title, 25) },
            { label: 'Tools', render: r => String(r.tool_count) },
            { label: 'Rewrite', render: r => String(r.rewrite_count) },
            { label: '模式', render: r => r.trace?.response_mode ? badge(r.trace.response_mode, 'info') : '-' },
            { label: '時間', render: r => formatDate(r.created_at) },
            { label: '操作', render: r => {
                const btn = el('button', { className: 'btn btn-sm', textContent: '詳情' });
                btn.addEventListener('click', () => showDetail(r.message_id));
                return btn;
            }},
        ], data.items));

        paginationContainer.appendChild(buildPagination(currentPage, data.total, limit, (p) => {
            currentPage = p;
            load();
        }));
    } catch (err) {
        showError(tableContainer, err.message);
    }
}

async function showDetail(messageId) {
    const { overlay, body } = createModal('Log 詳情');
    modalRoot.appendChild(overlay);
    showLoading(body);

    try {
        const data = await apiFetch(`/api/logs/${messageId}`);
        clear(body);

        body.appendChild(el('div', { className: 'detail-grid' },
            el('div', { className: 'detail-item' },
                el('div', { className: 'label', textContent: '使用者' }),
                el('div', { className: 'value', textContent: data.email || data.username }),
            ),
            el('div', { className: 'detail-item' },
                el('div', { className: 'label', textContent: '模式' }),
                el('div', { className: 'value', textContent: data.trace?.response_mode || '-' }),
            ),
            el('div', { className: 'detail-item' },
                el('div', { className: 'label', textContent: '執行時間' }),
                el('div', { className: 'value', textContent: data.trace?.total_execution_time ? `${data.trace.total_execution_time}s` : '-' }),
            ),
        ));

        if (data.trace?.tool_calls?.length) {
            body.appendChild(el('div', { className: 'card', style: 'margin-top:1rem' },
                el('div', { className: 'card-title', textContent: 'Tool 呼叫' }),
            ));
            data.trace.tool_calls.forEach(tc => {
                body.lastElementChild.appendChild(el('div', { className: 'trace-block' },
                    el('div', { className: 'trace-title', textContent: `${tc.node} → ${tc.name}` }),
                    el('div', { textContent: `Query: ${tc.query || '-'}` }),
                    tc.start_date ? el('div', { textContent: `日期: ${tc.start_date} ~ ${tc.end_date || ''}` }) : null,
                    tc.raw_args ? el('div', {}, el('div', { className: 'label', textContent: 'Raw Args' }), renderJson(tc.raw_args)) : null,
                ));
            });
        }

        if (data.trace?.query_rewrites?.length) {
            body.appendChild(el('div', { className: 'card', style: 'margin-top:1rem' },
                el('div', { className: 'card-title', textContent: 'Query Rewrite' }),
            ));
            data.trace.query_rewrites.forEach(rw => {
                body.lastElementChild.appendChild(el('div', { className: 'trace-block' },
                    el('div', { className: 'trace-title', textContent: `${rw.node} · ${rw.type}` }),
                    rw.pattern ? el('div', { textContent: `Pattern: ${rw.pattern}` }) : null,
                    rw.original_query ? el('div', { textContent: `原始: ${rw.original_query}` }) : null,
                    el('div', { textContent: `改寫: ${rw.rewritten_query || '-'}` }),
                ));
            });
        }

        if (data.token_logs?.length) {
            body.appendChild(el('div', { className: 'card', style: 'margin-top:1rem' },
                el('div', { className: 'card-title', textContent: 'Token 紀錄' }),
                buildTable([
                    { label: 'Caller', key: 'caller' },
                    { label: 'Model', key: 'model_name' },
                    { label: 'Tokens', render: r => String(r.total_tokens) },
                    { label: 'Cost', render: r => '$' + Number(r.cost_usd).toFixed(4) },
                ], data.token_logs),
            ));
        }

        if (data.trace?.steps?.length) {
            body.appendChild(el('div', { className: 'card', style: 'margin-top:1rem' },
                el('div', { className: 'card-title', textContent: '完整 Steps' }),
                renderJson(data.trace.steps),
            ));
        }
    } catch (err) {
        showError(body, err.message);
    }

    if (window.lucide) lucide.createIcons();
}

load();
